### PR TITLE
Update trimerge array, object, and map to handle internal equality

### PR DESCRIPTION
**Pull request by marcello3d**
With this change, when trimerging an object, array, or map, if the final
merged object/array/map has the same keys and values as left or right,
one of those two will be returned.

This will obviate the need to have a deep equality merger in every merge
chain, though you still want the simple `trimergeEquality` merger.

## Changes
- split up `trimerge-json.ts` into `trimerge-array.ts` and `trimerge-object.ts`
- simplify `trimergeArrayCreator` to not create an intermediate object, instead merge individual items as needed
- simplify `trimergeEquality` (rather than using creator)
- update `trimergeArrayCreator`, `trimergeObject`, and `trimergeMapCreator` to shallowly compare their merged values and return the original `left` or `right` structures if appropriate
- update tests to use reference equality checking when possible

### CODE REVIEW

File: src/js-same-type.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { CannotMerge } from './cannot-merge';
+import { type } from './type';
+
+export function jsSameType(
+  orig: any,
+  left: any,
+  right: any,
+): typeof orig | 'array' | 'null' | typeof CannotMerge {
```
Comment (srubin) [inline src/js-same-type.ts]: `typeof orig` here is `any`, which makes the entire return type the same as `any`.

File: src/trimerge-array.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    right: any[],
+    path: Path,
+    mergeFn: MergeFn,
+  ): JSONValue[] | typeof CannotMerge => {
+    if (jsSameType(orig, left, right) !== 'array') {
+      return CannotMerge;
+    }
+    const origMap: JSONObject = {};
+    const leftMap: JSONObject = {};
+    const rightMap: JSONObject = {};
```
Comment (srubin) [inline src/trimerge-array.ts]: We don't know for sure that `orig`/`left`/`right` are arrays of `JSONObject`s. Should these types be `Record<string, any>`?

File: src/trimerge-array.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      origKeys,
+      leftKeys,
+      rightKeys,
+      (key) => {
+        const leftElement = leftMap[key];
+        const rightElement = rightMap[key];
+        const merged = mergeFn(
+          origMap[key],
+          leftElement,
+          rightElement,
```
Comment (srubin) [inline src/trimerge-array.ts]: Could get `leftElement` and `rightElement` inline here.

File: src/trimerge-array.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+          [...path, key],
+          mergeFn,
+        );
+        if (merged !== left[mergedArray.length]) {
+          leftSame = false;
+        }
+        if (merged !== right[mergedArray.length]) {
+          rightSame = false;
+        }
```
Comment (srubin) [inline src/trimerge-array.ts]: Add a comment here explaining why these are comparing the merged value to the value at a specific index in `left` and `right`.

Comment (srubin) [reply src/trimerge-array.ts]: Also, micro-optimization: could add `leftSame &&` and `rightSame &&` to the beginning of these conditions.

File: src/trimerge-object.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  }
+  const origKeys = Object.keys(orig);
+  const leftKeys = Object.keys(left);
+  const rightKeys = Object.keys(right);
+  const mergedObject: JSONObject = {};
+  const keys = new Set<string>(origKeys);
+  leftKeys.forEach(keys.add, keys);
+  rightKeys.forEach(keys.add, keys);
+  let leftSame = true; //leftKeys.length === keys.size;
+  let rightSame = true; //rightKeys.length === keys.size;
```
Comment (srubin) [inline src/trimerge-object.ts]: Delete comments? Or make them more verbose :-)

File: src/trimerge-array.ts (hunk 4 above)
Comment (marcello3d) [reply src/trimerge-array.ts]: I had that originally but wasn’t sure if it would be faster

**Pull request merged**